### ISSUE TITLE
Java 9 Annotation Processing Broken

### ISSUE THREAD

**Issue by gudenau** [rzwitserloot/lombok#1472]
I am using Lombok 1.16.18 on JDK 9 and IntelliJ Idea 2017.2.4. 

When using Lombok's annotation processor it causes the compiler to throw an exception.
`Warning:(8, 8) java: lombok.javac.apt.LombokProcessor could not be initialized. Lombok will not run during this compilation: java.lang.IllegalArgumentException: com.sun.tools.javac.api.ClientCodeWrapper$WrappedStandardJavaFileManager extends com.sun.tools.javac.api.ClientCodeWrapper$WrappedJavaFileManager implements javax.tools.StandardJavaFileManager
  	at lombok.javac.apt.LombokFileObjects.getCompiler(LombokFileObjects.java:130)
  	at lombok.javac.apt.InterceptingJavaFileManager.<init>(InterceptingJavaFileManager.java:40)
  	at lombok.javac.apt.LombokProcessor.placePostCompileAndDontMakeForceRoundDummiesHook(LombokProcessor.java:164)
  	at lombok.javac.apt.LombokProcessor.init(LombokProcessor.java:85)
  	at lombok.core.AnnotationProcessor$JavacDescriptor.want(AnnotationProcessor.java:87)
  	at lombok.core.AnnotationProcessor.init(AnnotationProcessor.java:140)
  	at lombok.launch.AnnotationProcessorHider$AnnotationProcessor.init(AnnotationProcessor.java:69)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$ProcessorState.<init>(JavacProcessingEnvironment.java:675)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$DiscoveredProcessors$ProcessorStateIterator.next(JavacProcessingEnvironment.java:774)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:869)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.access$2200(JavacProcessingEnvironment.java:108)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1206)
  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1315)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1246)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:922)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
  	at org.jetbrains.jps.javac.JavacMain.compile(JavacMain.java:190)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.compileJava(JavaBuilder.java:473)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.compile(JavaBuilder.java:328)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.doBuild(JavaBuilder.java:255)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.build(JavaBuilder.java:208)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runModuleLevelBuilders(IncProjectBuilder.java:1260)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:937)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1009)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:900)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:733)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:385)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:192)
  	at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:138)
  	at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:295)
  	at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:125)
  	at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler.lambda$channelRead0$0(BuildMain.java:236)
  	at org.jetbrains.jps.service.impl.SharedThreadPoolImpl.lambda$executeOnPooledThread$0(SharedThreadPoolImpl.java:42)
  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
  	at java.base/java.lang.Thread.run(Thread.java:844)
`

**Comment by gudenau** [comment]
Side note, I never realized how much I used Lombok until it broke. Most of my classes need changing to compile correctly without it!

**Comment by Mart-Bogdan** [comment]
Sadly it's using internal api, hope it would be fixed in few monthes, thankfully Java8 is still ok :-)

**Comment by wrprice** [comment]
I'm also getting what appears to be the same error, but using the command line and _not_ IntelliJ.

    * Lombok edge `1.16.19 (2017-07-17 20:26:39 UTC)`
    * JDK 9 GA `Java HotSpot(TM) 64-Bit Server VM (build 9+181, mixed mode)`
    * Gradle 4.2 `2017-09-20 14:48:23 UTC`

I've opened the relevant modules/packages, and even tried `--illegal-access=permit`

```
warning: lombok.javac.apt.LombokProcessor could not be initialized. Lombok will not run during this compilation: java.lang.IllegalArgumentException: com.sun.tools.javac.api.ClientCodeWrapper$WrappedStandardJavaFileManager extends com.sun.tools.javac.api.ClientCodeWrapper$WrappedJavaFileManager implements javax.tools.StandardJavaFileManager
        at lombok.javac.apt.LombokFileObjects.getCompiler(LombokFileObjects.java:143)
        at lombok.javac.apt.InterceptingJavaFileManager.<init>(InterceptingJavaFileManager.java:40)
        at lombok.javac.apt.LombokProcessor.placePostCompileAndDontMakeForceRoundDummiesHook(LombokProcessor.java:164)
        at lombok.javac.apt.LombokProcessor.init(LombokProcessor.java:85)
```

**Comment by wrprice** [comment]
I've got a bit of a hack workaround locally.  Working on something I could submit as a starting point.  (Need to go through a contribution process w/ my employer since it's on work time.)

**Comment by tmurakam** [comment]
I've created the patch to fix this issue.
The wrappedManager seems to be created with wrapWithCallDispatcher() of https://github.com/JetBrains/intellij-community/blob/master/jps/jps-builders-6/src/org/jetbrains/jps/javac/JavacMain.java
and it does not extends BaseFileManager.

**Comment by gudenau** [comment]
Any news on when this will be on Maven?

**Comment by rspilker** [comment]
We've created an [edge release](https://projectlombok.org/download-edge). Can someone verify that this version fixes the problem?

**Comment by rspilker** [comment]
@gudenau We will make a new release after we get feedback on the edge release.

**Comment by krzyk** [comment]
Does it exist in any maven repo, or should I install it manually?

**Comment by krzyk** [comment]
@rspilker I can confirm that with edge release in IntelliJ IDEA 2013.3 EAP (Build #IU-173.3302.5, built on October 17, 2017) lombok annotations work fine.

**Comment by rspilker** [comment]
Thanks

**Comment by thetric** [comment]
The edge release unfortunately doesn't fix the Gradle build issue. The build still fails with the original stacktrace. (Built with Gradle 4.2.1, Java 9.0.1)

**Comment by wrprice** [comment]
+1 the new edge release works with Gradle 4.2.1

**Comment by krzyk** [comment]
There is a link on lombok download page: https://projectlombok.org/download-edge

**Comment by maystrovyy** [comment]
@krzyk supra, thank you!

**Comment by gudenau** [comment]
Interesting, I am getting an error regarding a javax error from modules not being visible.

```
Error:(17, 1) java: package javax.annotation is not visible
  (package javax.annotation is declared in module java.xml.ws.annotation, which is not in the module graph)

```
And the line with the error is
```
@ToString
```
on a class definition.

**Comment by thetric** [comment]
To fix this add this (assuming you're using gradle)

```gradle
compileOnly 'javax.annotation:javax.annotation-api:1.3.1'
```

**Comment by gudenau** [comment]
I am using intelliJ Idea, without a build system.

**Comment by rspilker** [comment]
To get rid of the javax.annotation problem, you could also create a lombok.config file, or [download this one](https://raw.githubusercontent.com/rzwitserloot/lombok/master/docker/shared/lombok.config).

**Comment by gudenau** [comment]
So the latest edge build does work for me.

**Comment by rspilker** [comment]
@gudenau Can you tell me what version of IntelliJ you've used, so I can update the compatibility notes?

**Comment by lpandzic** [comment]
I confirm it works with IntelliJ version 2017.2.

**Comment by gudenau** [comment]
IntelliJ Ultimate 2017.X, not sure what the latest is. I keep it up to date.

**Comment by kamilgregorczyk** [comment]
When can we expect it to be in maven?

**Comment by wrprice** [comment]
@rspilker - My organization is eagerly waiting for 1.16.20 since we need the fixes in the current edge release but our policy prohibits depending upon what is effectively a snapshot.  Since .18 is now four months old, is there anything else we can do to help unblock the release of .20?

**Comment by sskorol** [comment]
@wrprice could you please share your build.gradle configuration, which works for you with edge version?

**Comment by wrprice** [comment]
@sskorol - I can't share my entire project config, but these are the things I changed when I was initially trying to get it working:

```
dependencies {
    compileOnly files("path/to/lombok-edge.jar")
}

tasks.withType(JavaCompile) { // see note below
    options.fork = true
    options.forkOptions.jvmArgs += [
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED',
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED',
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED',
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED',
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED',
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED',
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED',
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED',
          '--add-opens', 'jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED'
        ]
}
```

As a test, I just **commented out** the `options.fork` and `options.forkOptions.jvmArgs` and my build is still working on JDK 9 GA and Gradle 4.2.1 using the newer `java-library` plugin instead of the older `java` plugin.  (*shrug*)  So, with that combination of software, it seems to be simply a matter of applying the new Lombok jar.

**Comment by krzyk** [comment]
@rspilker Are more tests needed to confirm the fix for IntelliJ? Or there are other issues that are targeted for .20 release? We are eagerly awaiting new release so we can use Java 9 in projects with lombok.

**Comment by rspilker** [comment]
We're still finalizing jdk9 support.

**Comment by slnowak** [comment]
@rspilker can we expect it this year?

**Comment by aalmazzo** [comment]
+1 please support java9

**Comment by jayantbit** [comment]
+1   
Lombok edge jar is working perfectly with gradle 4.2.1.
Please release it to maven central.

**Comment by rspilker** [comment]
1.16.20 has just been uploaded to our website and should be available on maven central soon.

**Comment by gudenau** [comment]
Cheers!

**Comment by lpandzic** [comment]
@rspilker, why was out of the box support for ConstructorProperties removed? It was working fine on edge release.

**Comment by krzyk** [comment]
I think lack of ConstructorProperties breaks behavior with deserialization with jackson. Is there a switch to bring it back?

**Comment by lpandzic** [comment]
Yes, this is exactly the issue I had, single parameter constructor doesn't work without an annotation.

**Comment by krzyk** [comment]
@lpandzic I found in the changelog that now we need to use `lombok.anyConstructor.addConstructorProperties = true` to have it work (and probably add java.desktop module)

**Comment by rspilker** [comment]
The reason it was removed by default is that it would break jdk9 compilation if you didn't specify the dependency on the extra module.

**Comment by randakar** [comment]
Damn. Would be nice to have that documented somewhere. I want to upgrade our projects to JDK 9 but gotcha's like that (we use Jackson, ofc) would make it more difficult. A small migration guide would help.

**Comment by krzyk** [comment]
@randakar https://projectlombok.org/changelog is not enough?

**Comment by lpandzic** [comment]
Well, for jackson, this was a hack anyway, we need real support like I've requested in this issue - https://github.com/FasterXML/jackson-databind/issues/1498.
Thanks @rspilker .

**Comment by randakar** [comment]
@krzyk I guess you do have a point there. I was thinking along the lines of documenting what happens if you upgrade an existing JDK 8 / jackson project, but if it's just this one single thing I guess the changelog will have to do.

**Comment by Arpit0492** [comment]
This worked for me https://github.com/mplushnikov/lombok-intellij-plugin/issues/415

**Comment by eximius313** [comment]
[This comment](https://github.com/rzwitserloot/lombok/issues/1472#issuecomment-343969972) could be related with [this bug](https://github.com/gradle/gradle/issues/3035)
